### PR TITLE
Register some backwards functions as operators

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44052 Register some backwards functions as operators**
* #44051 BatchedFallback: stop emitting the entire schema in the fallback warning

Summary
=======

This PR registers the following backwards functions as operators:
- slice_backward
- select_backward
- gather_backward
- index_select_backward (the backward function for index_select)
- value_selecting_reduction_backward (prevously known as index_select_backward, but is actually the backward function for max.dim, min.dim, etc)

In the future, I'd like to register more backward functions as operators
so that we can write batching rules for the backward functions. Batching
rules for backward functions makes it so that we can compute batched
gradients.

Motivation
==========
The rationale behind this PR is that a lot of backwards functions (27 in total)
are incompatible with BatchedTensor due to using in-place operations.
Sometimes we can allow the in-place operations, but other times we can't.
For example, consider select_backward:

```
Tensor select_backward(const Tensor& grad, IntArrayRef input_sizes, int64_t dim, int64_t index) {
  auto grad_input = at::zeros(input_sizes, grad.options());
  grad_input.select(dim, index).copy_(grad);
  return grad_input;
}
```

and consider the following code:
```
x = torch.randn(5, requires_grad=True)
def select_grad(v):
    torch.autograd.grad(x[0], x, v)

vs = torch.randn(B0)
batched_grads = vmap(select_grad)(vs)
```

For the batched gradient use case, `grad` is a BatchedTensor.
The physical version of `grad` has size `(B0,)`.
However, select_backward creates a `grad_input` of shape `(5)`, and
tries to copy `grad` to a slice of it.

Other approaches
================

I've considered the following:
- register select_backward as an operator (this PR)
- have a branch inside select_backward for if `grad` is batched.
    - this is OK, but what if we have more tensor extensions that want to override this?
- modify select_backward to work with BatchedTensor, by creating a new operator for the "select + copy_ behavior".
    - select + copy_ isn't used elsewhere in derivative formulas so this doesn't seem useful

Test Plan
=========

- `pytest test/test_autograd.py -v`
- Registering backward functions may impact performance. I benchmarked
select_backward to see if registering it as an operator led to any noticable
performance overheads: https://gist.github.com/zou3519/56d6cb53775649047b0e66de6f0007dc.
The TL;DR is that the overhead is pretty minimal.

Differential Revision: [D23481183](https://our.internmc.facebook.com/intern/diff/D23481183)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Thanks, you reasoning is well motivated. Be aware that functions with `_backward` in their name get treated specially in Python level codegen; make sure this is the behavior you want!

**Comment by zou3519** [comment]
> Thanks, you reasoning is well motivated. Be aware that functions with `_backward` in their name get treated specially in Python level codegen; make sure this is the behavior you want!

Yes, it looks like functions with `_backward` in their name don't get exposed to Python, which is exactly what I wanted (these backward operators should be "private" APIs)

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -576,7 +576,7 @@
   self: non_differentiable
 
 - name: kthvalue(Tensor self, int k, int dim=-1, bool keepdim=False) -> (Tensor values, Tensor indices)
-  self: index_select_backward(grad, dim, indices, self.sizes(), keepdim)
+  self: select_index_backward(grad, dim, indices, self.sizes(), keepdim)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:579]: These namings are quite confusing... Should we use something else now that they go into the aten API?
Like "reduction_dim_backward()"?

**Review by albanD — APPROVED** [review]
LGTM

minor comment about the naming as "index_select_backward" vs "select_index_backward" can be a bit confusing haha

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (zou3519) [reply tools/autograd/derivatives.yaml:579]: `select_index_backwards` can't be use for all reductions (like mean / sum). Is there some standard name for max(), topk(), mode(), median(), max(), mean()? All of those operations perform a reduction that selects specific values.

maybe "value_selecting_reduction_backward"?

**Comment by zou3519** [comment]
@albanD, @ezyang -- it looks like this change (registering backward functions as operators) has some implications for selective build. E.g. "select_backward" now needs to be included for a project that has selective builds and computes the backward of select. That doesn't seem like a problem, but let me know if you think this is not good

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (zou3519) [reply tools/autograd/derivatives.yaml]: Changed to "value_selecting_reduction_backward".

**Comment by ezyang** [comment]
@zou3519 Indeed. This should only matter for federated learning at the moment, @ljk53 I don't think we compute dependencies between forwards and backwards, but this seems like a logical thing to do when we need this eventually?
